compose/foundation/foundation/src/skikoMain/kotlin/androidx/compose/foundation/text/TextFieldDelegate.skiko.kt

```diff
@@ -0,0 +1,223 @@
+    val isGreaterThanFirst = caretOffset >= 0
+    val isLessThanLast = caretOffset <= currentText.lastIndex
+    return char.isPunctuationOrWhitespace() && isGreaterThanFirst && isLessThanLast
+}
+
+private fun TextLayoutResultProxy.isFirstLetterOfWordTapped(caretOffset: Int): Boolean {
+    return value.getWordBoundary(caretOffset).start == caretOffset
+}
+
+private fun isCaretTapped(
+    caretOffset: Int,
+    previousCaretOffset: Int
+): Boolean {
+    return previousCaretOffset == caretOffset
+}
+
+private fun TextLayoutResultProxy.isLeftEdgeTapped(caretOffset: Int): Boolean {
+    val lineNumber = value.getLineForOffset(caretOffset)
+    val lineStartOffset = value.getLineStart(lineNumber)
+    return lineStartOffset == caretOffset
+}
+
+private fun TextLayoutResultProxy.isRightEdgeTapped(caretOffset: Int): Boolean {
+    val lineNumber = value.getLineForOffset(caretOffset)
+    val lineEndOffset = value.getLineEnd(lineNumber)
+    return lineEndOffset == caretOffset
+}
+
+/**
+ * Recursively finds the next word boundary position starting from the given caret offset in the current text.
+ * The main difference between this and **getWordBoundary()** in *Paragraph.kt* is that
+ * this method finds next word boundary in case if caret positioned in between two words
+ * (so caret is pointing at punctuation or whitespace), whereas **getWordBoundary()** will return
+ * empty TextRange with the beginning on caret position
+ *
+ * @param caretOffset the offset of the caret position
+ * @param currentText the current text
+ * @param textLayoutResult the TextLayoutResultProxy object that provides text layout information
+ * @return the TextRange representing the next word boundary position
+ */
+private tailrec fun findNextWordBoundary(
+    caretOffset: Int,
+    currentText: String,
+    textLayoutResult: TextLayoutResultProxy
+): TextRange {
+    val wordRange = textLayoutResult.value.getWordBoundary(caretOffset)
+    val currentChar = currentText[caretOffset]
+    val lastIndex = currentText.lastIndex
+    return if (caretOffset > lastIndex) {
+        TextRange(lastIndex)
+    } else if (currentChar.isWhitespace()) {
+        findNextWordBoundary(caretOffset + 1, currentText, textLayoutResult)
+    } else if (!currentChar.isPunctuation()) {
+        wordRange // not a whitespace or punctuation
+    } else {
+        getSymbolsOnlySubsequenceBoundaries(
+            caretOffset,
+            currentText
+        ) // may be null (example: aaa...)
+            ?: findNextWordBoundary(caretOffset + 1, currentText, textLayoutResult)
+    }
+}
+
+private fun getSymbolsOnlySubsequenceBoundaries(caretOffset: Int, currentText: String): TextRange? {
+    var leadOffset = caretOffset
+    var trailOffset = caretOffset
+    val lastIndex = currentText.lastIndex
+    while (!currentText[leadOffset].isWhitespace() && leadOffset >= 0) {
+        if (!currentText[leadOffset].isPunctuation()) {
+            return null
+        }
+        leadOffset -= 1
```

Comment:
Unicode support problem ⚠️ 

`1` is NOT a size of one character. You can use `String.codePoints`/`String.codePointAt`/`String.codePointBefore`/`CodePoint.charCount()` that I wrote previously to handle multi-char Unicode symbols.

There are a few more similar places in this file, please revisit it too.

Reply:
moved this logic into CharHelpers.skiko.kt in ui-text, rewrote all similar places with code points usage